test/unit/html_test.rb

```diff
@@ -0,0 +1,331 @@
+      { insert: "\n", attributes: { id: 'bingo' } }
+    ])
+
+    f = {
+      oembed: {
+        tag: 'iframe',
+        block_format: :div,
+        apply: ->(el, op) { el[:src] = op.value.dig(:oembed, :url) }
+      }
+    }
+
+    assert_equal '<div id="bingo"><iframe src="https://www.youtube.com/watch?v=fd8tya7Gmv8"/></div>', render_compact_html(d, formats: f)
+  end
+
+  it 'renders custom object insertions with no block format' do
+    d = RichText::Delta.new([
+      { insert: { oembed: { url: "https://www.youtube.com/watch?v=fd8tya7Gmv8" } } },
+      { insert: "\n", attributes: { id: 'bingo' } }
+    ])
+
+    f = {
+      oembed: {
+        tag: 'iframe',
+        block_format: false,
+        apply: ->(el, op) { el[:src] = op.value.dig(:oembed, :url) }
+      }
+    }
+
+    assert_equal '<iframe src="https://www.youtube.com/watch?v=fd8tya7Gmv8" id="bingo"/>', render_compact_html(d, formats: f)
+  end
+
+  it 'renders custom object insertions with a build function' do
+    d = RichText::Delta.new([
+      { insert: { image: { src: "https://placekitten.com/200/150", credit: 'cuteness' } } },
+      { insert: "\n", attributes: { id: 'bingo' } }
+    ])
+
+    f = {
+      image: {
+        tag: 'img',
+        block_format: false,
+        build: ->(el, op) {
+          el[:src] = op.value.dig(:image, :src)
+          el.wrap('<image/>')
+          el = el.parent
+          el.add_child("<cite>#{ op.value.dig(:image, :credit) }</cite>")
+          el
+        }
+      }
+    }
+
+    assert_equal '<image id="bingo"><img src="https://placekitten.com/200/150"/><cite>cuteness</cite></image>', render_compact_html(d, formats: f)
+  end
+
+  it 'renders custom object insertions into story flow' do
+    d = RichText::Delta.new([
+      { insert: 'before' },
+      { insert: "\n" },
+      { insert: { image: { src: 'https://placekitten.com/200/150', credit: 'cuteness' } } },
+      { insert: "\n" },
+      { insert: 'after' },
+      { insert: "\n" },
+    ])
+
+    f = {
+      image: {
+        tag: 'img',
+        block_format: false,
+        build: ->(el, op) {
+          el[:src] = op.value.dig(:image, :src)
+          el.wrap('<figure/>')
+          el = el.parent
+          el.add_child("<cite>#{ op.value.dig(:image, :credit) }</cite>")
+          el
+        }
+      }
+    }
+
+    assert_equal '<p>before</p><figure><img src="https://placekitten.com/200/150"/><cite>cuteness</cite></figure><p>after</p>', render_compact_html(d, formats: f)
+  end
+
+  it 'gracefully handles missing newline ends' do
+    d = RichText::Delta.new([
+      { insert: 'mali principii' },
+      { insert: "\n" },
+      { insert: 'malus finis', attributes: { invalid: true } }
+    ])
+    assert_equal '<p>mali principii</p><p>malus finis</p>', render_compact_html(d)
+  end
+
+  def render_compact_html(delta, options={})
+    RichText::HTML.render_xml(delta, options).inner_html(save_with: 0)
```

Comment:
What does `save_with: 0` do here? Change the way whitespace is handled? I can't find good docs on that option from Nokogiri.

Reply:
Yep, that mysterious param is apparently the Nokogiri squish option for rendering without any pretty formatting... I found it on a stack overflow post and it did what I wanted for the test suite string formatting. It's a terrible option name, right?